### PR TITLE
store: gc stale peer

### PR DESCRIPTION
**Pull request by BusyJay**
ref: https://github.com/pingcap/tikv/issues/805

@ngaut @siddontang @disksing @zhangjinpeng1987 PTAL

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -270,21 +269,36 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                           region_id,
-                           msg.get_from_peer(),
-                           msg.get_to_peer(),
-                           e);
-                    return Err(e);
+        // we may encounter a message with larger peer id, which means
+        // current peer is stale, then we should remove current peer
+        let mut has_peer = false;
+        let mut stale_peer = None;
+        if let Some(p) = self.region_peers.get(&region_id) {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: should move this to a function

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -250,10 +250,47 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            has_peer = true;
+            let target_peer_id = target.get_id();
+            if p.peer_id() < target_peer_id {
+                if p.is_applying_snap() {
+                    // to remove the applying peer, we should find a reliable way to abort
+                    // the apply process.
+                    warn!("Stale peer {} is applying snapshot, will destroy next time.",
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: here means applying may still work?

Comment (BusyJay) [inline src/raftstore/store/store.rs]: Yes. As L263 explains.

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
PTAL @hhkbp2

**Comment by BusyJay** [comment]
ping

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**